src/v/bytes/details/io_fragment.h

```diff
@@ -94,14 +102,265 @@ class io_fragment {
+        using reference = value_type&;
+
+    private:
+        iter([[maybe_unused]] container_type* container, pointer curr)
+          : _current(curr)
+#ifndef NDEBUG
+          , _my_generation(container->_generation)
+          , _container(container)
+#endif
+        {
+        }
+
+    public:
+        iter() noexcept = default;
+        // Support conversion to const_iterator
+        // NOLINTNEXTLINE(hicpp-explicit-conversions)
+        operator iter<true, Forward>() const {
+            check_generation();
+#ifndef NDEBUG
+            return {_container, _current};
+#else
+            return {nullptr, _current};
+#endif
+        }
+
+        reference operator*() const {
+            check_generation();
+            return *_current;
+        }
+        pointer operator->() const {
+            check_generation();
+            return _current;
+        }
+        iter& operator++() {
+            check_generation();
+            if constexpr (Forward) {
+                _current = _current->_next;
+            } else {
+                _current = _current->_prev;
+            }
+            return *this;
+        }
+        iter operator++(int) {
+            auto tmp = *this;
+            ++*this;
+            return tmp;
+        }
+        bool operator==(const iter& o) const {
+            check_generation();
+            return _current == o._current;
+        }
+
+    private:
+        friend class io_fragment_list;
+
+        inline void check_generation() const {
+            dassert(
+              _container->_generation == _my_generation,
+              "Attempting to use an invalidated iterator. The corresponding "
+              "iobuf has been mutated since this iterator was constructed.");
+        }
+
+        pointer _current = nullptr;
+#ifndef NDEBUG
+        size_t _my_generation = 0;
+        container_type* _container = nullptr;
+#endif
+    };
+
+public:
+    io_fragment_list() noexcept = default;
+    io_fragment_list(io_fragment_list&& o) noexcept
+      : _head(std::exchange(o._head, nullptr))
+      , _tail(std::exchange(o._tail, nullptr)) {}
+    io_fragment_list(const io_fragment_list&) = delete;
+    io_fragment_list& operator=(io_fragment_list&&) = delete;
+    io_fragment_list& operator=(const io_fragment_list&) = delete;
+    ~io_fragment_list() noexcept {
+        dassert(
+          _head == nullptr && _tail == nullptr,
+          "io_fragment_list cleanup failure and possible memory leak");
+    };
+
+    using iterator = iter<false, true>;
+    using reverse_iterator = iter<false, false>;
+    using const_iterator = iter<true, true>;
+
+    bool empty() const {
+        check_consistency();
+        return _head == nullptr;
+    }
+
+    io_fragment& front() {
+        check_consistency();
+        return *_head;
+    }
+
+    void pop_front() {
+        check_consistency();
+        io_fragment* h = _head;
+        _head = h->_next;
+        h->_next = nullptr;
```

Comment:
Just curious, do we actually need to set _next of an element that we're removing? Same with pop_back()'s prev

Reply:
Only because we assert that we're cleaning up the pointer correctly (see the io_fragment destructor). We could technically not do that, but this makes it clearer that this node is "detached".